KAFKA-12887 Skip some RuntimeExceptions from exception handler

jlprat opened the pull request:
Instead of letting all `RuntimeException`s go through and be processed by
the uncaught exception handler, `IllegalStateException` and
`IllegalArgumentException` are not passed through and fail fast.

Added test checking this new case.
For the test I use an existing test as a baseline that was checking the uncaught exception handler was called and checked that it wasn't called for this new particular case.

Possible extensions would be to add more types of `RuntimeException`s on the new `catch` clause.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jlprat committed:
KAFKA-12887 Skip some RuntimeExceptions from exception handler

Instead of letting all RuntimeExceptions go through and be processed by
the uncaught exception handler, IllegalStateException and
IllegalArgumentException are not passed through and fail fast.

Added test checking this new case.

jlprat commented:
cc. @guozhangwang Feel free to review this PR and let me know if this is what you had in mind with https://issues.apache.org/jira/browse/KAFKA-12887

jlprat commented:
Test failure was https://issues.apache.org/jira/browse/KAFKA-8529 which was apparently resolved. I reopened the issue.
I rebased this PR to the trunk of a couple of hours ago.

jlprat commented:
Reopening to re-trigger CI

guozhangwang reviewed (commented):
Thanks @jlprat The proposed PR looks promising to me. Though regarding how to handle illegal state / argument, I was thinking a bit more aggressive: right now by throwing the exception directly we just let that thread being killed but still letting the other threads proceed, i.e. the instance would still be alive with one thread less. I'm thinking we should just kill the whole instance via "SHUTDOWN_CLIENT" (this would require us to add another private function like `handleStreamsUncaughtException(StreamThreadExceptionResponse)`.

Also cc @cadonna who implemented the original cases here. LMK WDYT.

jlprat commented:
Thanks for your feedback @guozhangwang 
I thought "simply" killing the thread was the original intention.
I'll look into killing the whole instance once @cadona provides his feedback on this (just in case there isn't something easy to overlook)

jlprat committed:
Remove previous implementation

jlprat committed:
Shut down client when certain runtime exceptions are thrown

When setting the uncaught exception handler we check if the exception is in an "exclude list", if so, we terminate the client, otherwise we continue as usual.

Adds integration test checking that user defined exception handler is not used when an IllegalStateException is thrown.

jlprat commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -495,9 +498,25 @@ private void replaceStreamThread(final Throwable throwable) {
+    private StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse getActionForThrowable(final Throwable throwable,
+                                                                                        final StreamsUncaughtExceptionHandler streamsUncaughtExceptionHandler) {
+        final StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse action;
+        // Exception might we wrapped within a StreamsException one
```
https://github.com/jlprat/kafka/blob/KAFKA-12887-no-trigger-exceptionhandler-for-rte/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L756
This is where I can see RTE are always wrapped in a `StreamsException`. I'm not sure if this is always happening or there are cases in which the RTE is not wrapped.

guozhangwang replied:
Hmm I think it should be "always" wrapped, if not then that would be a bug.. before line 500 we can check if `cause()` exists and if that's illegal-state etc.

jlprat replied:
Updated now, the `if` statement now only checks the wrapped exception if it exists, and decides upon that. See next line

jlprat committed:
Throw another exception that is not IllegalStateException as now it shuts down the client

jlprat commented on streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java:
```diff
@@ -104,7 +104,7 @@ public void shouldEmitSameRecordAfterFailover() throws Exception {
             .map((key, value) -> {
                 if (shouldThrow.compareAndSet(true, false)) {
-                    throw new IllegalStateException("Kaboom");
+                    throw new RuntimeException("Kaboom");
```
Throwing an `IllegalStateException` would after this change, cause a client shut down.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -495,9 +498,25 @@ private void replaceStreamThread(final Throwable throwable) {
+            || (throwable.getCause() != null && EXCEPTIONS_NOT_TO_BE_HANDLED_BY_USERS.contains(throwable.getCause().getClass()))) {
+            // The exception should not be passed over to the user defined uncaught exception handler.
+            // Something unexpected happened, we should shut down the client
+            log.warn("Exception bypassed the user defined uncaught exception handler."); // No need to print the exception, it will be printed in the next log line.
```
nit: instead of adding a `warn` before the `error` below, I think it's sufficient to just change the error log (a.k.a.line 530) and remove `the registered exception opted to` instead just say `and Kafka Streams opted to`. Currently it would reads like, first "exception bypassed the handler" and then "handler opted to shutdown" which is self-conflicting.

guozhangwang reviewed (commented):
Just one minor comment, otherwise LGTM.

jlprat committed:
Apply feedback

jlprat commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -509,7 +526,7 @@ private void handleStreamsUncaughtException(final Throwable throwable,
             case SHUTDOWN_CLIENT:
                 log.error("Encountered the following exception during processing " +
-                        "and the registered exception handler opted to " + action + "." +
+                        "and Kafka Streams opted to " + action + "." +
```
Changed the error log line as mentioned.

jlprat reviewed (commented):
@guozhangwang applied your feedback in here. Feel free to review if you have some time, thanks!

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -495,9 +498,23 @@ private void replaceStreamThread(final Throwable throwable) {
         }
     }
 
+    private StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse getActionForThrowable(final Throwable throwable,
+                                                                                        final StreamsUncaughtExceptionHandler streamsUncaughtExceptionHandler) {
```
```suggestion
    private StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse getActionForThrowable(final Throwable throwable,
                                                                                                final StreamsUncaughtExceptionHandler streamsUncaughtExceptionHandler) {
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -495,9 +498,23 @@ private void replaceStreamThread(final Throwable throwable) {
+                                                                                        final StreamsUncaughtExceptionHandler streamsUncaughtExceptionHandler) {
+        final StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse action;
+        // Exception might we wrapped within a StreamsException one
+        if (throwable.getCause() != null && EXCEPTIONS_NOT_TO_BE_HANDLED_BY_USERS.contains(throwable.getCause().getClass())) {
```
Could you please put this check in a method with a meaningful name like `wrappedExceptionIsIn()`? Then the check would be 
```
if (wrappedExceptionIsIn(EXCEPTIONS_NOT_TO_BE_HANDLED_BY_USERS))
``` 
and you can remove all other inline comments because the code per se contains all needed info.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/StreamsUncaughtExceptionHandlerIntegrationTest.java:
```diff
@@ -163,6 +164,27 @@ public void shouldShutdownClient() throws InterruptedException {
 
+
+    @Test
+    public void shouldShutdownClientWhenIllegalStateException() throws InterruptedException {
```
Could you please also add a test for the `IllegalArgumentException`?

cadonna reviewed (commented):
Thanks for the updates, @jlprat !

Here my feedback!

jlprat committed:
Extract check to method and add new test

jlprat commented:
@cadonna Applied your feedback as well, thanks for the comments. I also realized the `AtomicBoolean` variables in the test could be final and added that as well.

jlprat committed:
Fix checkstyle rule

jlprat commented:
Test failure is related to https://issues.apache.org/jira/browse/KAFKA-8138 but not the same, as in this particular instance it seems more like a failure in the test itself (topic already exists)

jlprat commented:
Ping @cadonna and @guozhangwang Is there anything else you'd like changed in the PR?

fabian-wiesman commented:
Aiven Inc - https://aiven.io/ is misusing of Apache Kafka trademark as 'Aiven Kafka' [1] and violates
Apache trademarks. See - [1] https://help.aiven.io/en/?q=%22aiven+kafka%22 This simply means Aiven Inc / contributors have no idea of Apache projects / communities or how open source works in general. CC @ivanyu , @hnousiainen, Kafka Community / Project Management Committee.

guozhangwang reviewed (approved):
LGTM!

guozhangwang commented:
Merged to trunk, thanks @jlprat !

jlprat commented:
Thanks for the reviews @guozhangwang and @cadonna !
